//===----------------------------------------------------------------------===//
//
// This source file is part of the Soto for AWS open source project
//
// Copyright (c) 2017-2020 the Soto project authors
// Licensed under Apache License v2.0
//
// See LICENSE.txt for license information
// See CONTRIBUTORS.txt for the list of Soto project authors
//
// SPDX-License-Identifier: Apache-2.0
//
//===----------------------------------------------------------------------===//

// THIS FILE IS AUTOMATICALLY GENERATED by https://github.com/swift-aws/soto/blob/main/CodeGenerator/Sources/CodeGenerator/main.swift. DO NOT EDIT.

import SotoCore

/// Error enum for PinpointEmail
public struct PinpointEmailErrorType: AWSErrorType {
    enum Code: String {
        case accountSuspendedException = "AccountSuspendedException"
        case alreadyExistsException = "AlreadyExistsException"
        case badRequestException = "BadRequestException"
        case concurrentModificationException = "ConcurrentModificationException"
        case limitExceededException = "LimitExceededException"
        case mailFromDomainNotVerifiedException = "MailFromDomainNotVerifiedException"
        case messageRejected = "MessageRejected"
        case notFoundException = "NotFoundException"
        case sendingPausedException = "SendingPausedException"
        case tooManyRequestsException = "TooManyRequestsException"
    }

    private let error: Code
    public let context: AWSErrorContext?

    /// initialize PinpointEmail
    public init?(errorCode: String, context: AWSErrorContext) {
        guard let error = Code(rawValue: errorCode) else { return nil }
        self.error = error
        self.context = context
    }

    internal init(_ error: Code) {
        self.error = error
        self.context = nil
    }

    /// return error code string
    public var errorCode: String { self.error.rawValue }

    public static var accountSuspendedException: Self { .init(.accountSuspendedException) }
    public static var alreadyExistsException: Self { .init(.alreadyExistsException) }
    public static var badRequestException: Self { .init(.badRequestException) }
    public static var concurrentModificationException: Self { .init(.concurrentModificationException) }
    public static var limitExceededException: Self { .init(.limitExceededException) }
    public static var mailFromDomainNotVerifiedException: Self { .init(.mailFromDomainNotVerifiedException) }
    public static var messageRejected: Self { .init(.messageRejected) }
    public static var notFoundException: Self { .init(.notFoundException) }
    public static var sendingPausedException: Self { .init(.sendingPausedException) }
    public static var tooManyRequestsException: Self { .init(.tooManyRequestsException) }
}

extension PinpointEmailErrorType: Equatable {
    public static func == (lhs: PinpointEmailErrorType, rhs: PinpointEmailErrorType) -> Bool {
        lhs.error == rhs.error
    }
}

extension PinpointEmailErrorType: CustomStringConvertible {
    public var description: String {
        return "\(self.error.rawValue): \(self.message ?? "")"
    }
}
